Implement chain follower for the stake pool metrics

Anviking opened the pull request:
# Issue Number

#713 


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I added a simple `worker` that uses existing functionality to collect, merge and write stake-pool metrics to the DB.
- [x] Added a proper `Block`-type instead of `(BlockHeader, Maybe PoolId)` 

# Comments

- Note: I'm too uncomfortable with the `WorkerRegistry`, so I'm not using it now. 
- Testing: 🤔 Maybe:
    - killing the worker and re-restarting doesn't affect the results (using a mock network layer)



<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Stub worker for storing stake distrubution in DB

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -52,8 +73,105 @@ import GHC.Generics
+        return Continue
+
+    retryPolicy = limitRetries 3
+    getHeader = header
```
? :thinking:

Anviking replied:
My thought was to highlight that it was a `x -> BlockHeader` and not a `BlockHeader`.

Can remove if too much indirection.

KtorZ replied:
I'd either rename the original accessor or leave it  as `header` (which is fine IMO, but that's a naming debate I don't want to go through right now :sweat_smile: ...)

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -185,14 +303,13 @@ instance Buildable State where
-applyBlock (newTip, poolId) (State _prevTip prevMap) =
-        State newTip (Map.alter alter poolId prevMap)
+applyBlock :: Block -> State -> State
+applyBlock (Block newTip mpoolId) s@(State _prevTip prevMap) =
```
are we still using this :thinking: ?

Anviking replied:
No. I could remove in a commit in this PR.

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -52,8 +73,105 @@ import GHC.Generics
+        -- If the tip is still the same after we have fetched the stake
+        -- distribution, we conclude that the stake-distrubtion is from the same
+        -- chain as the tip.
+        consistentDistr <- withinSameTip retryPolicy liftE getTip $ \nodeTip2 -> do
```
what about using guards here?

```hs
\nodeTip2
    | nodeTip /= nodeTip2 -> throwE ErrMetricsWorkeTipTooVolatile 
    | otherwise -> do
         ....
```

KtorZ replied:
Although, we probably want to retry here... Perhaps it's best to amend `withinSameTip` to that it takes an initial tip as argument!

Anviking replied:
Good point… But actually… Maybe the whole retrying in `withinSameTip` is pointless.

We care about `stakeDistribution nl` being called while the the tip is still `nodeTip` (from the follow-callback). It should be enough to:
```haskell
res <- action
after <- getTip
guard (after == nodeTip)
``` 
i.e not fetching an extra tip before running the action. 

Furthermore: if the comparison fails, retrying won't help. (unless the node switches from A to B and back to A again, which we already assumed it cannot)

Anviking replied:
8c04a4d

Anviking replied:
But _maybe_ we should then also
- Remove the possibility of `Retry` in `FollowAction` (because we will never retry)
- Consider allowing callbacks of `[]` such that we can fetch stake distribution even if there are no new blocks. What if the node has stopped syncing with the network? We should still be able to fetch the stake-distribution.

Anviking commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -52,8 +73,105 @@ import GHC.Generics
+            -> Retry
+        Left e
+            -> ExitWith e
+            -- TODO: Make all cases explicit here.
```
Not 100% if this indirection is good or if `advance'` should be merged into `advance`.

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -52,8 +73,105 @@ import GHC.Generics
+            case producer block of
+                Just pool ->
+                    -- TODO: What do we do if the production already exists?!
+                    withExceptT (const ErrMetricsWorkerDataInconsistencyDuplicateBlock) $
```
We can't recover from that (it means that either our roll back mechanism or chain follower is broken). This is a 500 and a valid case for Exception.

Anviking replied:
Note that the `worker` (collecting metrics and writing to DB) will now be completely separate from a future `StakePoolLayer` (reading from DB).

I guess we should recover by clearing the `pool_production` table in the DB.

KtorZ replied:
Hmmm. No. This will silently swallow a data inconsistency issue that shouldn't have happened. If this happens, we must crash.

Anviking replied:
>  we must crash

Crash what exactly? The worker thread? The entire wallet? (we would need added logic for the latter I believe)

Anviking replied:
We could also return `ExitOn ErrMetricsWorkerDataInconsistencyDuplicateBlock` to the callback such that it stops.

--> 0083e5c

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -52,8 +73,105 @@ import GHC.Generics
+            -> Continue
+
+    advance
+        :: NE.NonEmpty Block
```
The prefix is superfluous / inadequate here. We use qualified imports to disambiguate ambiguous functions like `map`, `insert` or `union` which requires a context. We avoid them for types.

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -52,8 +73,105 @@ import GHC.Generics
+        liftIO $ rollbackTo db point
+        return Continue
+
+    retryPolicy = limitRetries 3
```
This would read better inlined, it is quite far from where it's used

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -52,8 +73,105 @@ import GHC.Generics
+                    -- here would be /very/ unexpected.
+                    --
+                    -- TODO: How do we handle this?
+                    throwE ErrMetricsWorkerDataInconsistencyBlockWithoutProducer
```
If we expect all blocks to have a producer here, then we should have the `Block` type to reflect that :+1:

Anviking replied:
Would have to be something like `Either Genesis Block` in that case. Maybe that's good.

Anviking replied:
Right, of course `Either Genesis Block` doesn't fix the problem. It just re-packages the `Nothing` case as a `Genesis`.

We would have to parameterise `NetworkLayer` by both `block0` and `block`… Since our problem is that the conversion-function only works for non-genesis blocks. 🤔

Anviking replied:
Specialised `block0` to `W.Block Tx` as discussed.

Anviking committed:
Remove now obselete in-memory couting of block-producers

Anviking committed:
Specialize the type of block0 in NetworkLayer

See comment:

Note: The genesis block is a Wallet @Block@-type and not the
@block@ type-parameter used in @nextBlocks@. This is due to:
1. "Cardano.Pool.Metrics" wanting to use a @block@-type specific to
non-genesis blocks.
2. The genesis block only being needed in the Wallet (currently).

Anviking committed:
Change producer from `Maybe PoolId` to `PoolId`

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -14,55 +14,155 @@
+    , producer :: PoolId
+    } deriving (Eq, Show, Generic)
+
+data ErrMetricsWorkerErr
```
Err...Err ?

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -14,55 +14,155 @@
+                (limitRetries 3)
+                (const $ throwE ErrMetricsWorkerErrTipTooVolatile)
+                getTip
+                $ \nodeTip2 -> do
```
suggestion:

```hs
let liftE = ...
consistentDist <- withinSameTip (limitRetries 3) liftE getTip $ ...
```

Also, it seems that we are simply swallowing any errors returned by the callback? `ErrMetricsWorkerStakeDistributionUnavailible` will never be returned.

Anviking replied:
Reverted to 
```haskell
          where
            liftE ErrWithinSameTipMaxRetries =
                throwE ErrMetricsWorkerErrTipTooVolatile

```

KtorZ replied:
Isn't this a partial pattern match though...? Ah, no, nevermind.

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -14,55 +14,155 @@
+
+data ErrMetricsWorkerErr
+    = ErrMetricsWorkerErrTipTooVolatile
+    | ErrMetricsWorkerStakeDistributionUnavailible
```
`Unavailible` --> `Unavailable`

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -14,55 +14,155 @@
+            -- TODO: We might want to batch the insertion to the DB.
+            -- TODO: What do we do if the production already exists?!
+            withExceptT
+                (const ErrMetricsWorkerDataInconsistencyDuplicateBlock)
```
If the production already exists, we should throw an exception here. That shouldn't be possible.

Anviking replied:
cf https://github.com/input-output-hk/cardano-wallet/pull/876#discussion_r337125548

Anviking committed:
Remove withinSameTip in favor of more direct stakeDistributionConsistentWithTip

Anviking committed:
ExitOn ErrMetricsWorkerDataInconsistencyDuplicateBlock

KtorZ committed:
review chain follower worker in the stake pool metrics

KtorZ commented on lib/core/src/Cardano/Pool/Metrics.hs:
```diff
@@ -14,60 +14,137 @@
+    initCursor = do
+        let (_block0, bp) = staticBlockchainParameters nl
+        let k = fromIntegral . getQuantity . view #getEpochStability $ bp
+        readCursor db k
```
Actually, something to be careful with: the cursor might be empty. In which case we should use the genesis hash as a starting point.

KtorZ committed:
Use the genesis block hash to initialize the metrics cursor if there's not yet any checkpoints
